File: chronicle/viewer/typed-json.js
"use strict";

/* A flat typed JSON graph used wherever JavaScript and Python must agree on
 * exact wire identity. Flat nodes make valid deeply nested approval detail a
 * data-size concern, never a language recursion-limit concern. */
(function (root, factory) {
  const api = factory();
  if (typeof module === "object" && module.exports) module.exports = api;
  else root.BurrowTypedJSON = api;
})(typeof globalThis === "object" ? globalThis : this, function () {
  function exactDenseArray(value) {
    if (!Array.isArray(value) || Object.getPrototypeOf(value) !== Array.prototype ||
        Object.getOwnPropertySymbols(value).length) return false;
    const names = Object.getOwnPropertyNames(value);
    if (names.length !== value.length + 1 || names.at(-1) !== "length") return false;
    for (let index = 0; index < value.length; index++) {
      if (names[index] !== String(index)) return false;
      const descriptor = Object.getOwnPropertyDescriptor(value, names[index]);
      if (!descriptor || !Object.hasOwn(descriptor, "value") || !descriptor.enumerable ||
          !descriptor.configurable || !descriptor.writable) return false;
    }
    const length = Object.getOwnPropertyDescriptor(value, "length");
    return Boolean(length && length.value === value.length && length.writable && !length.enumerable &&
      !length.configurable);
  }
  /* Direct-object callers do not get to smuggle behavior or hidden state into
   * the JSON domain. JSON.parse produces exactly this shape: an ordinary/null
   * prototype object whose own string names are all enumerable data fields. */
  function exactPlainObject(value) {
    if (value === null || typeof value !== "object" || Array.isArray(value) ||
        Object.getOwnPropertySymbols(value).length) return false;
    const prototype = Object.getPrototypeOf(value);
    if (prototype !== Object.prototype && prototype !== null) return false;
    const names = Object.getOwnPropertyNames(value);
    const keys = Object.keys(value);
    if (names.length !== keys.length || names.some((name, index) => name !== keys[index])) return false;
    return names.every(name => {
      const descriptor = Object.getOwnPropertyDescriptor(value, name);
      return Boolean(descriptor && descriptor.enumerable && Object.hasOwn(descriptor, "value"));
    });
  }
  function stringToken(value) {
    let encoded = '"';
    for (let index = 0; index < value.length; index++) {
      const code = value.charCodeAt(index);
      if (code === 0x22) encoded += '\\"';
      else if (code === 0x5c) encoded += "\\\\";
      else if (code === 0x08) encoded += "\\b";
      else if (code === 0x09) encoded += "\\t";
      else if (code === 0x0a) encoded += "\\n";
      else if (code === 0x0c) encoded += "\\f";
      else if (code === 0x0d) encoded += "\\r";
      else if (code >= 0x20 && code <= 0x7e) encoded += value[index];
      else encoded += "\\u" + code.toString(16).padStart(4, "0");
    }
    return encoded + '"';
  }

  function numberBits(value) {
    if (!Number.isFinite(value)) return "nonfinite";
    if (value === 0) return "0000000000000000";
    const bytes = new Uint8Array(8);
    new DataView(bytes.buffer).setFloat64(0, value, false);
    return [...bytes].map(byte => byte.toString(16).padStart(2, "0")).join("");
  }

  function typedGraph(value) {
    const nodes = [], seen = new Set(), active = new Set(), root = { index: null };
    const stack = [{ value, target: root, key: "index", exit: false }];
    while (stack.length) {
      const frame = stack.pop(), item = frame.value;
      const array = Array.isArray(item);
      const object = item !== null && typeof item === "object" && !array;
      if (!array && !object) {
        let node;
        if (item === null) node = ["n"];
        else if (typeof item === "boolean") node = ["b", item ? 1 : 0];
        else if (typeof item === "number") node = ["f", numberBits(item)];
        else if (typeof item === "string") node = ["s", item];
        else node = ["x"];
        nodes.push(node); frame.target[frame.key] = nodes.length - 1; continue;
      }
      if (frame.exit) {
        active.delete(item);
        let node;
        if (array) node = ["a", frame.children];
        else node = ["o", frame.keys.map((key, index) => [key, frame.children[index]])];
        nodes.push(node); frame.target[frame.key] = nodes.length - 1; continue;
      }
      // JSON wire values are trees. A repeated container identity is therefore
      // a hostile direct-object graph even when it is not an active cycle. If
      // aliases were expanded independently, a tiny DAG could amplify before
      // the encoded capsule byte ceiling is reached.
      if (seen.has(item)) throw new TypeError(active.has(item) ?
        "cyclic JSON value" : "aliased JSON value");
      seen.add(item);
      if ((!array && !exactPlainObject(item)) || (array && !exactDenseArray(item))) {
        nodes.push(["x"]); frame.target[frame.key] = nodes.length - 1; continue;
      }
      const keys = array ? null : Object.keys(item).sort((left, right) => {
        const a = stringToken(left), b = stringToken(right);
        return a < b ? -1 : a > b ? 1 : 0;
      });
      let children;
      if (array) {
        children = new Array(item.length);
        // This is a trust boundary. Do not consult an input-owned/inherited
        // iterator even after proving the ordinary dense-array shape.
        for (let index = 0; index < item.length; index++) children[index] = item[index];
      } else children = keys.map(key => item[key]);
      const childIndexes = new Array(children.length);
      active.add(item);
      stack.push({ ...frame, exit: true, keys, children: childIndexes });
      for (let index = children.length - 1; index >= 0; index--) {
        stack.push({ value: children[index], target: childIndexes, key: index, exit: false });
      }
    }
    return [nodes, root.index];
  }

  function graphString(graph) {
    const nodes = graph[0], root = graph[1];
    const encoded = new Array(nodes.length);
    for (let position = 0; position < nodes.length; position++) {
      const node = nodes[position];
      const tag = node[0];
      if (tag === "n" || tag === "x") encoded[position] = `["${tag}"]`;
      else if (tag === "b") encoded[position] = `["b",${node[1]}]`;
      else if (tag === "f" || tag === "s") encoded[position] =
        `["${tag}",${stringToken(node[1])}]`;
      else if (tag === "a") encoded[position] = `["a",[${node[1].join(",")}]]`;
      else {
        const entries = new Array(node[1].length);
        for (let index = 0; index < node[1].length; index++) {
          entries[index] = `[${stringToken(node[1][index][0])},${node[1][index][1]}]`;
        }
        encoded[position] = `["o",[${entries.join(",")}]]`;
      }
    }
    return `[[${encoded.join(",")}],${root}]`;
  }

  function identity(value) { return graphString(typedGraph(value)); }

  function decodeGraph(graph) {
    if (!exactDenseArray(graph) || graph.length !== 2 || !exactDenseArray(graph[0]) ||
        !Number.isSafeInteger(graph[1])) throw new TypeError("invalid typed JSON graph");
    const values = [], references = new Array(graph[0].length).fill(0);
    for (let position = 0; position < graph[0].length; position++) {
      const node = graph[0][position];
      if (!exactDenseArray(node) || typeof node[0] !== "string") throw new TypeError("invalid typed JSON node");
      const tag = node[0]; let value;
      if (tag === "n" && node.length === 1) value = null;
      else if (tag === "b" && node.length === 2 && (node[1] === 0 || node[1] === 1)) value = Boolean(node[1]);
      else if (tag === "s" && node.length === 2 && typeof node[1] === "string") value = node[1];
      else if (tag === "f" && node.length === 2 && typeof node[1] === "string") {
        if (node[1] === "nonfinite") value = Infinity;
        else {
          if (!/^[0-9a-f]{16}$/.test(node[1])) throw new TypeError("invalid binary64 token");
          const bytes = Uint8Array.from(node[1].match(/../g), byte => Number.parseInt(byte, 16));
          value = new DataView(bytes.buffer).getFloat64(0, false);
          if (!Number.isFinite(value) || (value === 0 && node[1] !== "0000000000000000")) {
            throw new TypeError("noncanonical binary64 token");
          }
        }
      } else if (tag === "a" && node.length === 2 && exactDenseArray(node[1])) {
        value = new Array(node[1].length);
        for (let index = 0; index < node[1].length; index++) {
          const reference = node[1][index];
          if (!Number.isSafeInteger(reference) || reference < 0 || reference >= position) {
            throw new TypeError("invalid typed JSON array reference");
          }
          references[reference]++;
          value[index] = values[reference];
        }
      } else if (tag === "o" && node.length === 2 && exactDenseArray(node[1])) {
        value = {};
        let prior = null;
        for (let index = 0; index < node[1].length; index++) {
          const entry = node[1][index];
          if (!exactDenseArray(entry) || entry.length !== 2 || typeof entry[0] !== "string" ||
              !Number.isSafeInteger(entry[1]) || entry[1] < 0 || entry[1] >= position ||
              (prior !== null && TYPED_COMPARE(prior, entry[0]) >= 0)) {
            throw new TypeError("invalid typed JSON object entry");
          }
          references[entry[1]]++;
          Object.defineProperty(value, entry[0], { value: values[entry[1]], enumerable: true,
            configurable: true, writable: true });
          prior = entry[0];
        }
      } else throw new TypeError("invalid typed JSON node");
      values.push(value);
    }
    if (graph[1] < 0 || graph[1] >= values.length || graph[1] !== values.length - 1) {
      throw new TypeError("invalid typed JSON root");
    }
    if (references.some((count, index) =>
      index === graph[1] ? count !== 0 : count !== 1)) {
      throw new TypeError("noncanonical typed JSON tree");
    }
    const value = values[graph[1]];
    if (graphString(typedGraph(value)) !== graphString(graph)) {
      throw new TypeError("noncanonical typed JSON graph");
    }
    return value;
  }

  const TYPED_COMPARE = (left, right) => {
    const a = stringToken(left), b = stringToken(right);
    return a < b ? -1 : a > b ? 1 : 0;
  };

  return { exactDenseArray, exactPlainObject, stringToken, numberBits, typedGraph,
    graphString, identity, decodeGraph };
});

File: chronicle/typed_json.py
"""Exact identity for JSON values: canonical escaping and a typed binary64 graph.

Nothing here knows about approvals, events or HTTP.  The module answers one
question — *are these two JSON values the same value?* — in a domain where
``json.dumps`` is not an answer: it renders ``1`` and ``1.0`` differently while
treating ``0.1 + 0.2`` and ``0.30000000000000004`` as the same text, it leans on
dict ordering, and its escaping varies with keyword arguments.  Here numbers are
compared as the exact IEEE-754 binary64 the browser and the server both hold,
object keys are sorted, and every string is escaped one way.

**The public interface**

- ``canonical_string(value)`` — a JSON string literal, ASCII-only, one spelling.
- ``typed_graph(value)`` / ``decode_graph(graph)`` — a value as a flat, tagged,
  child-before-parent node list, and back.
- ``semantic_key(value)`` — ``typed_graph`` rendered to one comparable string.
- ``freeze_json(value)`` / ``thaw_json(value)`` — detached immutable containers,
  and ordinary ones again for an I/O boundary.

**The byte obligation**

This encoding is a wire and storage format, not an implementation detail, so the
bytes are part of the interface:

``retention._encode_mood_authority`` concatenates ``canonical_string`` and
``semantic_key`` output into a ``typed-binary64-v1`` capsule, ``event_log``
writes that line into the rotated log and its archives, and the next boot reads
it back through ``decode_graph``.  A capsule that fails to decode is dropped
silently, taking Mood authority history with it — so a change to the key sort
order or to a binary64 token orphans records already on disk, with no error to
notice.  Every other caller compares values in memory and is free to change.

Until 2026-08 the format also had to match a JavaScript twin, ``typed-json.js``,
byte for byte.  That file and the whole browser viewer were deleted in
warren#219, so the obligation is now to *stored data* rather than to a second
implementation; the surviving parity vectors in
``tests/fixtures/mood-capsule-parity.json`` are re-pointed at this codec alone.
``tests/test_typed_json.py`` pins the format with golden strings — control
characters, non-BMP surrogate pairs, binary64 tokens and a mood-authority
record.  Changing one of those is a log-format migration, not a refactor.

Durable *notification* identity is a separate encoder living in
``notification_persistence``; it depends on the approval shape, not on these
bytes.  Do not assume a change here is safe there, or the reverse.

**The hardening**

``decode_graph`` takes bytes off disk and, historically, off the wire, so it
validates before it restores: no forward or repeated references, no unsorted or
duplicated object keys, no non-canonical binary64 token, no orphan nodes, and a
final re-encode that must reproduce the input exactly.  Both directions run
iteratively — approval detail thousands of levels deep must not exhaust the
recursion stack in a server that has to stay up.
"""

from collections.abc import Mapping
import math
import re
import struct
from types import MappingProxyType


def _transform_json(value, freeze):
    """Transform JSON containers without consuming Python's recursion stack."""
    results = {}
    seen = set()
    active = set()
    stack = [(value, False)]
    while stack:
        item, exiting = stack.pop()
        if not isinstance(item, (Mapping, list, tuple)):
            results[id(item)] = item
            continue
        identity = id(item)
        if exiting:
            active.remove(identity)
            if isinstance(item, Mapping):
                mapped = {key: results[id(child)] for key, child in item.items()}
                results[identity] = MappingProxyType(mapped) if freeze else mapped
            else:
                values = [results[id(child)] for child in item]
                results[identity] = tuple(values) if freeze else values
            continue
        if identity in seen:
            raise ValueError(
                "cyclic JSON value" if identity in active else "aliased JSON value"
            )
        seen.add(identity)
        active.add(identity)
        stack.append((item, True))
        children = (
            [
                item[key]
                for key in sorted(item, key=lambda key: canonical_string(str(key)))
            ]
            if isinstance(item, Mapping)
            else list(item)
        )
        stack.extend((child, False) for child in reversed(children))
    return results[id(value)]


def freeze_json(value):
    """Return an immutable, iteratively detached representation of JSON data."""
    return _transform_json(value, True)


def thaw_json(value):
    """Return ordinary JSON containers for serialization at an I/O boundary."""
    return _transform_json(value, False)


def canonical_string(value):
    """Return ``value`` as a canonical compact ASCII JSON string literal.

    One spelling per string: short escapes for the five that have them, literal
    printable ASCII, and lowercase ``\\uXXXX`` for everything else — including
    DEL and every non-BMP character, which is emitted as its UTF-16 surrogate
    pair.  Unpaired surrogates pass through as their own code unit rather than
    raising, because Python strings decoded with ``surrogatepass`` carry them.
    """
    encoded = ['"']
    short = {8: "\\b", 9: "\\t", 10: "\\n", 12: "\\f", 13: "\\r"}
    for character in value:
        codepoint = ord(character)
        units = (
            [codepoint]
            if codepoint <= 0xFFFF
            else [
                0xD800 + ((codepoint - 0x10000) >> 10),
                0xDC00 + ((codepoint - 0x10000) & 0x3FF),
            ]
        )
        for code in units:
            if code == 34:
                encoded.append('\\"')
            elif code == 92:
                encoded.append("\\\\")
            elif code in short:
                encoded.append(short[code])
            elif 0x20 <= code <= 0x7E:
                encoded.append(chr(code))
            else:
                encoded.append(f"\\u{code:04x}")
    return "".join(encoded) + '"'


def typed_graph(value):
    """Flat, unambiguous JSON AST with exact normalized binary64 numbers.

    Children always precede their parent, so both encoding and decoding remain
    iterative even when valid public approval detail is thousands of levels deep.
    """
    nodes = []
    seen = set()
    active = set()
    root = [None]
    stack = [(value, root, 0, False, None, None)]
    while stack:
        item, target, target_index, exiting, keys, child_indexes = stack.pop()
        container = isinstance(item, (Mapping, list, tuple))
        if not container:
            if item is None:
                node = ["n"]
            elif type(item) is bool:
                node = ["b", 1 if item else 0]
            elif type(item) in (int, float):
                try:
                    number = float(item)
                except (OverflowError, ValueError):
                    number = math.inf
                bits = (
                    (
                        "0000000000000000"
                        if number == 0
                        else struct.pack(">d", number).hex()
                    )
                    if math.isfinite(number)
                    else "nonfinite"
                )
                node = ["f", bits]
            elif isinstance(item, str):
                node = ["s", item]
            else:
                node = ["x"]
            nodes.append(node)
            target[target_index] = len(nodes) - 1
            continue
        identity = id(item)
        if exiting:
            active.remove(identity)
            if isinstance(item, Mapping):
                node = [
                    "o",
                    [
                        [str(key), child_indexes[index]]
                        for index, key in enumerate(keys)
                    ],
                ]
            else:
                node = ["a", child_indexes]
            nodes.append(node)
            target[target_index] = len(nodes) - 1
            continue
        if identity in seen:
            raise ValueError(
                "cyclic JSON value" if identity in active else "aliased JSON value"
            )
        seen.add(identity)
        keys = (
            sorted(item, key=lambda key: canonical_string(str(key)))
            if isinstance(item, Mapping)
            else None
        )
        children = (
            [item[key] for key in keys] if isinstance(item, Mapping) else list(item)
        )
        child_indexes = [None] * len(children)
        active.add(identity)
        stack.append((item, target, target_index, True, keys, child_indexes))
        for index in range(len(children) - 1, -1, -1):
            stack.append((children[index], child_indexes, index, False, None, None))
    return [nodes, root[0]]


def _graph_json(graph):
    """Canonical shallow JSON for a typed graph."""

    def scalar(value):
        if isinstance(value, str):
            return canonical_string(value)
        return str(value)

    nodes, root = graph
    encoded_nodes = []
    for node in nodes:
        tag = node[0]
        if tag in {"n", "x"}:
            encoded_nodes.append('["' + tag + '"]')
        elif tag in {"b", "f", "s"}:
            encoded_nodes.append('["' + tag + '",' + scalar(node[1]) + "]")
        elif tag == "a":
            encoded_nodes.append('["a",[' + ",".join(map(str, node[1])) + "]]")
        else:
            entries = [
                "[" + canonical_string(key) + "," + str(index) + "]"
                for key, index in node[1]
            ]
            encoded_nodes.append('["o",[' + ",".join(entries) + "]]")
    return "[[" + ",".join(encoded_nodes) + "]," + str(root) + "]"


def decode_graph(graph):
    """Validate and iteratively restore a value produced by ``typed_graph``."""
    if (
        not isinstance(graph, list)
        or len(graph) != 2
        or not isinstance(graph[0], list)
        or type(graph[1]) is not int
    ):
        raise ValueError("invalid typed JSON graph")
    values = []
    references = [0] * len(graph[0])
    for position, node in enumerate(graph[0]):
        if not isinstance(node, list) or not node or not isinstance(node[0], str):
            raise ValueError("invalid typed JSON node")
        tag = node[0]
        if tag == "n" and len(node) == 1:
            value = None
        elif (
            tag == "b" and len(node) == 2 and type(node[1]) is int and node[1] in (0, 1)
        ):
            value = bool(node[1])
        elif tag == "s" and len(node) == 2 and isinstance(node[1], str):
            value = node[1]
        elif tag == "f" and len(node) == 2 and isinstance(node[1], str):
            token = node[1]
            if token == "nonfinite":
                value = math.inf
            else:
                if re.fullmatch(r"[0-9a-f]{16}", token) is None:
                    raise ValueError("invalid binary64 token")
                value = struct.unpack(">d", bytes.fromhex(token))[0]
                if not math.isfinite(value) or (
                    value == 0 and token != "0000000000000000"
                ):
                    raise ValueError("noncanonical binary64 token")
                if value.is_integer() and abs(value) <= 9007199254740991:
                    value = int(value)
        elif tag == "a" and len(node) == 2 and isinstance(node[1], list):
            if any(
                type(index) is not int or not 0 <= index < position for index in node[1]
            ):
                raise ValueError("invalid typed JSON array reference")
            for index in node[1]:
                references[index] += 1
            value = [values[index] for index in node[1]]
        elif tag == "o" and len(node) == 2 and isinstance(node[1], list):
            value = {}
            prior = None
            for entry in node[1]:
                if (
                    not isinstance(entry, list)
                    or len(entry) != 2
                    or not isinstance(entry[0], str)
                    or type(entry[1]) is not int
                    or not 0 <= entry[1] < position
                    or (
                        prior is not None
                        and canonical_string(prior) >= canonical_string(entry[0])
                    )
                ):
                    raise ValueError("invalid typed JSON object entry")
                references[entry[1]] += 1
                value[entry[0]] = values[entry[1]]
                prior = entry[0]
        else:
            raise ValueError("invalid typed JSON node")
        values.append(value)
    if graph[1] < 0 or graph[1] >= len(values) or graph[1] != len(values) - 1:
        raise ValueError("invalid typed JSON root")
    if any(
        count != (0 if index == graph[1] else 1)
        for index, count in enumerate(references)
    ):
        raise ValueError("noncanonical typed JSON tree")
    value = values[graph[1]]
    if _graph_json(typed_graph(value)) != _graph_json(graph):
        raise ValueError("noncanonical typed JSON graph")
    return value


def semantic_key(value):
    """Hashable exact identity in the browser's IEEE-754 consumer domain."""
    return _graph_json(typed_graph(value))
